src/Build/Globbing/MSBuildGlob.cs

```diff
@@ -0,0 +1,271 @@
+        //{
+        //    return _expandedFiles.Value;
+        //}
+
+        /// <summary>
+        ///     Similar to <see cref="IsMatch" /> but also provides the match groups for the glob parts
+        /// </summary>
+        /// <param name="stringToMatch"></param>
+        /// <returns></returns>
+        public MatchInfoResult MatchInfo(string stringToMatch)
+        {
+            ErrorUtilities.VerifyThrowArgumentNull(stringToMatch, nameof(stringToMatch));
+
+            if (FileUtilities.PathIsInvalid(stringToMatch) ||
+                !IsLegal)
+            {
+                return MatchInfoResult.Empty;
+            }
+
+            var normalizedInput = NormalizeMatchInput(stringToMatch);
+
+            bool isMatch;
+            string fixedDirectoryPart, wildcardDirectoryPart, filenamePart;
+            FileMatcher.GetRegexMatchInfo(
+                normalizedInput,
+                _regex.Value,
+                out isMatch,
+                out fixedDirectoryPart,
+                out wildcardDirectoryPart,
+                out filenamePart);
+
+            return new MatchInfoResult(isMatch, fixedDirectoryPart, wildcardDirectoryPart, filenamePart);
+        }
+
+        private string NormalizeMatchInput(string stringToMatch)
+        {
+            var rootedInput = Path.Combine(_globRoot, stringToMatch);
+            var normalizedInput = FileUtilities.GetFullPathNoThrow(rootedInput);
+
+            // Degenerate case when the string to match is empty.
+            // Ensure trailing slash because the fixed directory part has a trailing slash.
+            if (stringToMatch == string.Empty)
+            {
+                normalizedInput = normalizedInput + Path.DirectorySeparatorChar;
+            }
+
+            return normalizedInput;
+        }
+
+        /// <summary>
+        ///     Parse the given <paramref name="fileSpec" /> into a <see cref="MsBuildGlob" /> using a given
+        ///     <paramref name="globRoot" />.
+        /// </summary>
+        /// <param name="globRoot">
+        ///     The root of the glob.
+        ///     The fixed directory part of the glob and the match arguments (<see cref="IsMatch" /> and <see cref="MatchInfo" />)
+        ///     will get normalized against this root.
+        ///     If empty, the current working directory is used.
+        ///     Cannot be null, and cannot contain invalid path arguments.
+        /// </param>
+        /// <param name="fileSpec">The string to parse</param>
+        /// <returns></returns>
+        public static MsBuildGlob Parse(string globRoot, string fileSpec)
+        {
+            ErrorUtilities.VerifyThrowArgumentNull(globRoot, nameof(globRoot));
+            ErrorUtilities.VerifyThrowArgumentNull(fileSpec, nameof(fileSpec));
+            ErrorUtilities.VerifyThrowArgumentInvalidPath(globRoot, nameof(globRoot));
+
+            if (globRoot == string.Empty)
+            {
+                globRoot = Directory.GetCurrentDirectory();
```

Comment:
The only current driver of the API so far is GetAllGlobs, which gets them from ItemSpecs. ItemSpecs will always pass in the path of their project file as the glob root. Except for in-memory projects, which will have an empty path. Should I add an optional parameter for the current directory?